File: python/sort/merge.sort.py
arr = [20,40,30,100,5023,409,48,4348,394]



print("Original array is: ",arr)

def mergeSort(left:list[int],right:list[int])->list[int]:
    result:list[int] = []
    i:int = 0
    j:int = 0
    
    while i < len(left) and j < len(right):
        result.append(left[i] if left[i] < right[j] else right[j])
        if left[i] < right[j]:
            i += 1
        else:
            j += 1

            
    result.extend(left[i:])
    result.extend(right[j:])
            
    return result

def merge(arr:list[int]) -> list[int]:
    if len(arr) < 2:
        return arr
    
    mid:int = len(arr) // 2
    left:list[int] = arr[:mid]
    right:list[int] = arr[mid:]
    
    
    return mergeSort(merge(left),merge(right))


result = merge(arr)

print("Sorted array is: ",result)

File: js/Sort/merge.sort.js
const arr = [20, 40, 220,502, 3059, 503, 54, 250, 849, 0o0];

console.log("Original array is: ",arr)

function merge(arr){
    if(arr.length < 2){
        return arr;
    }


    let mid = Math.floor(arr.length / 2);
    let left = arr.slice(0,mid);
    let right = arr.slice(mid,arr.length);


    return mergeSort(merge(left),merge(right))
}


function mergeSort(left,right){
    let result = [];
    let leftIndex = 0;
    let rightIndex = 0;

    while(left.length > leftIndex && right.length > rightIndex){
        if(left[leftIndex] < right[rightIndex]){
            result.push(left[leftIndex++])
        }else{
            result.push(right[rightIndex++])
        }
    }

    return [...result,...left.slice(leftIndex),...right.slice(rightIndex)]
}

const sorted = merge(arr)

console.log("Sorted array is: ",sorted)